From 41c410f21252d58bb1cfc64a2df51cda91497923 Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Wed, 15 Oct 2025 16:47:31 +0200
Subject: [PATCH 422/553] CMake: Relax handling of CMP0156 policy

Don't warn about CMP0156 policy values for platforms we don't require
them for, which are Apple and Emscripten. This avoids scaring project
developers unnecessarily and forcing them to set the policy to OLD /
NEW explicitly.

Effectively this means we don't force the policy anymore to OLD if
cmake_minimum_required is 3.29 or higher on non-Apple and
non-Emscripten platforms.

CMake is smart enough to do de-duplication of libraries on platforms
or linkers that don't support it.

Amends c20d7bcb86361d0c9f8af3807dcad9db1a3a5ca0
Amends d6d832d0d28ac29807a04f6634a20e12649b440d

[ChangeLog][Build Systems] CMake user projects will now use CMake's
default value for the CMP0156 policy, except for Apple and Emscripten
platforms which are forced to NEW (when the policy is available).

Pick-to: 6.8
Fixes: QTBUG-141181
Task-number: QTBUG-135978
Task-number: QTBUG-140211
Change-Id: I44d65537a6ca38fdc16bf42088e18f2690f0adbf
Reviewed-by: Joerg Bornemann <joerg.bornemann@qt.io>
(cherry picked from commit b2402684cc18db631c41383293cb84bb2351d9d2)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/QtPublicCMakeVersionHelpers.cmake | 36 +++++++++++++------------
 1 file changed, 19 insertions(+), 17 deletions(-)

diff --git a/cmake/QtPublicCMakeVersionHelpers.cmake b/cmake/QtPublicCMakeVersionHelpers.cmake
index 292d97c84e7..0207c087e1a 100644
--- a/cmake/QtPublicCMakeVersionHelpers.cmake
+++ b/cmake/QtPublicCMakeVersionHelpers.cmake
@@ -107,10 +107,9 @@ endfunction()
 # Handle force-assignment of CMP0156 policy when using CMake 3.29+.
 #
 # For Apple-platforms we set it to NEW, to avoid duplicate linker issues when using -ObjC flag.
+# For Emscripten / WebAssembly we also set it to NEW, to avoid duplicate linker issues.
 #
-# For non-Apple platforms we set it to OLD, because we haven't done the necessary testing to
-# see which platforms / linkers can handle the new deduplication behavior, without breaking the
-# various linking techniques that Qt uses for object library propagation.
+# For other platforms, we leave the policy value as-is, without showing any warnings.
 function(__qt_internal_set_cmp0156)
     # Exit early if not using CMake 3.29+
     if(NOT POLICY CMP0156)
@@ -164,29 +163,32 @@ function(__qt_internal_set_cmp0156)
         set(default_policy_value NEW)
         set(unsupported_policy_value OLD)
     else()
-        # For non-Apple linkers, we keep the previous behavior of not deduplicating libraries,
-        # because we haven't done the necessary testing to identify on which platforms
-        # it is safe to deduplicate.
-        set(default_policy_value OLD)
-        set(unsupported_policy_value NEW)
+        # For other platforms we don't enforce any policy values and keep them as-is.
+        set(default_policy_value "")
+        set(unsupported_policy_value "")
     endif()
 
     # Force set the default policy value for the given platform, even if the policy value is
     # the same or empty. That's because in the calling function scope, the value can be empty
     # due to the cmake_minimum_required call in Qt6Config.cmake resetting the policy value.
-    get_cmake_property(debug_message_shown _qt_internal_cmp0156_debug_message_shown)
-    if(NOT debug_message_shown)
-        message(DEBUG "Force setting the CMP0156 policy to '${default_policy_value}' "
-            "for platform '${CMAKE_SYSTEM_NAME}'.")
-        set_property(GLOBAL PROPERTY _qt_internal_cmp0156_debug_message_shown TRUE)
-    endif()
+    if(default_policy_value)
+        get_cmake_property(debug_message_shown _qt_internal_cmp0156_debug_message_shown)
+        if(NOT debug_message_shown)
+            message(DEBUG "Force setting the CMP0156 policy to '${default_policy_value}' "
+                "for platform '${CMAKE_SYSTEM_NAME}'.")
+            set_property(GLOBAL PROPERTY _qt_internal_cmp0156_debug_message_shown TRUE)
+        endif()
 
-    cmake_policy(SET CMP0156 "${default_policy_value}")
+        cmake_policy(SET CMP0156 "${default_policy_value}")
+    endif()
 
-    # If the policy is explicitly set to a value other than the default, issue a warning.
+    # If the policy is explicitly set to a value other than the (non-empty) default, issue a
+    # warning.
     # Don't show the warning if the policy is unset, which would be the default for most
     # projects, because it's too much noise. Also don't show it for Qt builds.
-    if("${policy_value}" STREQUAL "${unsupported_policy_value}" AND NOT QT_BUILDING_QT)
+    if(unsupported_policy_value
+            AND "${policy_value}" STREQUAL "${unsupported_policy_value}"
+            AND NOT QT_BUILDING_QT)
         message(WARNING
             "CMP0156 is set to '${policy_value}'. Qt forces the '${default_policy_value}'"
             " behavior of this policy for the '${CMAKE_SYSTEM_NAME}' platform by default."
-- 
2.51.2

